Rakefile

```
@@ -33,13 +33,58 @@ namespace :resque do
 end
 
 require 'rake/testtask'
+
 Rake::TestTask.new do |t|
   $stdout.sync = true
   $stderr.sync = true
   ENV['APP_ENV'] ||= 'test'
   t.pattern = 'test/**/*_test.rb'
 end
 
+Rake::TestTask.new(:test_structure) do |t|
+  $stdout.sync = true
+  $stderr.sync = true
+  ENV['APP_ENV'] ||= 'test'
+  t.test_files = ['test/api/**/*_test.rb', 'test/models/**/*_test.rb']
+end
+
+namespace :test do
+  task :api do
+    ENV['COV'] = 'false'
+    ENV['TEST'] ||= 'test/api/**/*_test.rb'
+    Rake::Task['test'].invoke
+  end
+
+  task :models do
+    ENV['COV'] = 'false'
+    ENV['TEST'] ||= 'test/models/**/*_test.rb'
+    Rake::Task['test'].invoke
+  end
+
+  task :structure do
+    ENV['COV'] = 'false'
+    Rake::Task['test_structure'].invoke
+  end
+
+  task :lib do
+    ENV['COV'] = 'false'
+    ENV['TEST'] ||= 'test/lib/**/*_test.rb'
+    Rake::Task['test'].invoke
+  end
+
+  task :clustering do
+    ENV['COV'] = 'false'
+    ENV['TEST'] ||= 'test/**/*clustering*_test.rb'
+    Rake::Task['test'].invoke
+  end
+
+  task :periodic do
+    ENV['COV'] = 'false'
+    ENV['TEST'] ||= 'test/**/*scheduling*_test.rb'
+    Rake::Task['test'].invoke
+  end
+end
+
```

```suggestion
namespace :test do
  Rake::TestTask.new(:api){ |t| t.pattern = 'test/api/**/*_test.rb' }

  Rake::TestTask.new(:model){ |t| t.pattern = 'test/models/**/*_test.rb' }

  Rake::TestTask.new(:structure){ |t| t.test_files = ['test/api/**/*_test.rb', 'test/models/**/*_test.rb'] }

  Rake::TestTask.new(:lib){ |t| t.pattern = 'test/lib/**/*_test.rb' }

  Rake::TestTask.new(:clustering){ |t| t.pattern = 'test/**/*clustering*_test.rb' }

  Rake::TestTask.new(:periodic){ |t| t.pattern = 'test/**/*scheduling*_test.rb' }
end

```

This way `rake -T` shows the sub tasks correctly and we don't have to touch the `TEST` variable which stays free to be used injunction if necessary.

If we set the default value of `COV` correctly down below then we don't need to fiddle with it here.

---

About the **TEST** variable, these tasks are meant to launch a folder, I don't see the point of leaving the TEST variable free. However I prefer the `t.pattern` as it adds the description in `rake -T` command. 

Removing the `ENV['COV']` is not a good idea for me, I explain it below.

---

I think your argument "these tasks are meant to launch a folder" is incorrect because these tasks are meant to launch all the tests related to a "concept" and they are not necessarily in a single folder. 

For example, `:periodic` launches the following 4 tests
```
optimizer-api/test/real_cases_scheduling_solver_test.rb
optimizer-api/test/lib/heuristics/scheduling_functions_test.rb
optimizer-api/test/lib/heuristics/scheduling_instance_validity_test.rb
optimizer-api/test/lib/heuristics/scheduling_test.rb
```
but it doesn't launch the other tests in these folders -- `optimizer-api/test/lib/heuristics/`, `optimizer-api/test/`. 

I pointed out **TEST** because it can be used injunction to further **filter** the tests as follows: `rake test:periodic TEST="test/lib/heuristics/*"` would launch all the scheduling heuristic `library` tests but not the `real_cases` ones. For me leaving this variable free for the advanced use cases like this is a nice **cherry on the top**.

---

I don't understand why you suggest not touching `ENV['COV']` here, it is the opposite of your suggestion below. There you suggest that we shouldn't make the default behavior without coverage but here you suggest leaving `ENV['COV']='false'` in place... As I explained below, I think the code should do its job and the procedures should do their job, we shouldn't make the code less elegant and more complex to make it handle the duty of CI procedures.

---

did you test all tasks Hali with your suggestions?

---

:+1:  yeap they are all good